langchain/src/vectorstores/singlestore.ts

```diff
@@ -4,18 +4,81 @@ import type {
   OkPacket,
   ResultSetHeader,
   FieldPacket,
+  PoolOptions,
 } from "mysql2/promise";
 import { format } from "mysql2";
+import { createPool } from "mysql2/promise";
+import { Metadata } from "@opensearch-project/opensearch/api/types.js";
 import { VectorStore } from "./base.js";
 import { Embeddings } from "../embeddings/base.js";
 import { Document } from "../document.js";
 
-export interface SingleStoreVectorStoreConfig {
-  connectionPool: Pool;
+export type DistanceMetrics = "DOT_PRODUCT" | "EUCLIDEAN_DISTANCE";
+
+const OrderingDirective: Record<DistanceMetrics, string> = {
+  DOT_PRODUCT: "DESC",
+  EUCLIDEAN_DISTANCE: "",
+};
+
+export interface ConnectionOptions extends PoolOptions {}
+
+type ConnectionWithUri = {
+  connectionOptions?: never;
+  connectionURI: string;
+};
+
+type ConnectionWithOptions = {
+  connectionURI?: never;
+  connectionOptions: ConnectionOptions;
+};
+
+type ConnectionConfig = ConnectionWithUri | ConnectionWithOptions;
+
+export type SingleStoreVectorStoreConfig = ConnectionConfig & {
   tableName?: string;
   contentColumnName?: string;
   vectorColumnName?: string;
   metadataColumnName?: string;
+  distanceMetrics?: DistanceMetrics;
+};
+
+function withConnectAttributes(
+  config: SingleStoreVectorStoreConfig
+): ConnectionOptions {
+  let newOptions: ConnectionOptions = {};
+  if (config.connectionURI) {
+    newOptions = {
+      uri: config.connectionURI,
+    };
+  } else if (config.connectionOptions) {
+    newOptions = {
+      ...config.connectionOptions,
+    };
+  }
+  const result: ConnectionOptions = {
+    ...newOptions,
+    connectAttributes: {
+      ...newOptions.connectAttributes,
+    },
+  };
+
+  if (!result.connectAttributes) {
+    result.connectAttributes = {};
+  }
+
+  if (
+    !Object.prototype.hasOwnProperty.call(
```

Comment:
What does `Object.prototype.hasOwnProperty.call` do vs. just `result.connectAttributes.program_name !== undefined`?

Reply:
Would prefer the latter for readability